perf(transformer/object_rest_spread): use `ArenaVec` to store values that will be used in constructing AST

Dunqing opened the pull request:
`props` is always used in `expression_object` without intermediate changes, so it is okay and good to use `ArenaVec` store, which also can avoid an unnecessary allocation (std vec -> arena vec converison).

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10434?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10434** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10434?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10434?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10418** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10418?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#9881](https://github.com/oxc-project/oxc/pull/9881) <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9881?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10434).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
### Merge activity

* **Apr 16, 7:47 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
There's something cool we could do here with our new `Vec` implementation.

We could add an API `Vec::split_at` which converts a `Vec` into 2 x `Vec`s, without copying any data or making any allocations.

```rs
let original_len = vec.len();
let original_ptr = vec.as_ptr();

let (vec1, vec2) = vec.split_at(2);

assert!(vec1.len() == 2);
assert!(vec2.len() == original_len - 2);
// `vec1` and `vec2` point to chunks of the original `vec`
assert!(vec1.as_ptr() == original_ptr);
assert!(vec2.as_ptr() == original_ptr.add(2));
```

`std`'s `Vec` cannot offer this API because when a `Vec` is dropped, it must deallocate with the same layout as it was originally allocated with. But because our `Vec` is not `Drop`, we don't have that restriction.

Using `split_at`, this transform could avoid copying any of the data from original `Vec` into new `Vec`s. It'd just create a bunch of `Vec`s which reference chunks of the original's data.

Dunqing committed:
perf(transformer/object_rest_spread): use `ArenaVec` to store values that will be used in constructing AST (#10434)

`props` is always used in `expression_object` without intermediate changes, so it is okay and good to use `ArenaVec` store, which also can avoid an unnecessary allocation (std vec -> arena vec converison).

Dunqing commented:
> There's something cool we could do here with our new `Vec` implementation.
> 
> We could add an API `Vec::split_at` which converts a `Vec` into 2 x `Vec`s, without copying any data or making any allocations.
> 
> ```rust
> let original_len = vec.len();
> let original_ptr = vec.as_ptr();
> 
> let (vec1, vec2) = vec.split_at(2).unwrap();
> 
> assert!(vec1.len() == 2);
> assert!(vec2.len() == original_len - 2);
> // `vec1` and `vec2` point to chunks of the original `vec`
> assert!(vec1.as_ptr() == original_ptr);
> assert!(vec2.as_ptr() == original_ptr.add(2));
> ```
> 
> `std`'s `Vec` cannot offer this API because when a `Vec` is dropped, it must deallocate with the same layout as it was originally allocated with. But because our `Vec` is not `Drop`, we don't have that restriction.
> 
> Using `split_at`, this transform could avoid copying any of the data from original `Vec` into new `Vec`s. It'd just create a bunch of `Vec`s which reference chunks of the original's data.

That's cool! Maybe a dumb question, what scenario do you want to use this API for? In terms of the PR changes, we don't need to use `vec1`, so it looks like we should have a method like `cut_off`?

```rust
let original_vec = vec![1, 2, 3, 4, 5];
let cut_vec = vec.cut_off(2).unwrap();

assert!(original_vec.len() == 2);
assert!(cut_vec.len() == original_vec.len() - 2);
assert!(cut_vec.as_ptr() == original_ptr.add(2));
```
